File: .net standard2/Weed3/DataItem.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace Noear.Weed {
    public class DataItem  : IDataItem {
        Dictionary<String, Object> _data = new Dictionary<string, object>();

        public DataItem() { }
        public DataItem(bool isUsingDbNull) { _isUsingDbNull = isUsingDbNull; }

        public int count() {
            return _data.Count;
        }

        public void clear() {
            _data.Clear();
        }

        public bool exists(string name) {
            return _data.ContainsKey(name);
        }

        public IEnumerable<string> keys() {
            return _data.Keys;
        }
        
        public IDataItem set(String name, Object value) {
            _data[name] = value;
            return this;
        }

        public Object get(String name) {
            return _data[name];
        }

        public Variate getVariate(String name) {
            if (_data.ContainsKey(name))
                return new Variate(name, get(name));
            else
                return new Variate(name, null);
        }

        public T toItem<T>(T item) where T : IBinder {
            item.bind((key)=>{
                return getVariate(key);
            });

            return item;
        }

        public short getShort(String name) {
            return (short)_data[name];
        }

        public int getInt(String name) {
            return (int)_data[name];
        }

        public int getInt2(String name) {
            Object val = get(name);

            if (val == null) {
                return 0;
            }

            if (val is long){
                return (int)((long)val);
            }

            if (val is int){
                return (int)val;
            }

            if (val is Decimal){
                return (int)((Decimal)val);
            }

            return 0;
        }

        public long getLong(String name) {
            return (long)_data[name];
        }

        public long getLong2(String name) {
            Object val = get(name);

            if (val == null) {
                return 0;
            }

            if (val is long){
                return (long)val;
            }

            if (val is int){
                return (int)val;
            }

            if (val is Decimal){
                return (long)((Decimal)val);
            }

            return 0;
        }

        public double getDouble(String name) {
            return (double)_data[name];
        }

        public float getFloat(String name) {
            return (float)_data[name];
        }

        public String getString(String name) {
            return (String)_data[name];
        }

        public bool getBoolean(String name) {
            return (bool)_data[name];
        }

        public DateTime getDateTime(String name) {
            return (DateTime)_data[name];
        }

        public void forEach(Action<String, Object> callback) {
            foreach (var kv in _data) {
                if (kv.Value == null && _isUsingDbNull) {
                    callback(kv.Key, "$NULL");
                }
                else {
                    callback(kv.Key, kv.Value);
                }
            }
        }

        private bool _isUsingDbNull = false;

        //============================
        public static IDataItem create(IDataItem schema, GetHandler source) {
            DataItem item = new DataItem();
            foreach (var key in schema.keys()) {
                object val = source(key);
                if (val != null) {
                    item.set(key, val);
                }
            }
            return item;
        }
    }
}

File: weed3/src/main/java/org/noear/weed/DataItem.java
package org.noear.weed;

import org.noear.weed.ext.Act2;
import org.noear.weed.ext.Fun2;
import org.noear.weed.utils.EntityUtils;

import java.util.*;
import java.util.function.Consumer;

/**
 * Created by noear on 14-9-10.
 *
 * 不能转为继承自Map
 * 否则，嵌入别的引擎时，会变转为不可知的MapAdapter
 */
public class DataItem implements IDataItem, Iterable<Map.Entry<String,Object>>{
    LinkedHashMap<String,Object> _data = new LinkedHashMap<>();

    public DataItem() { }
    public DataItem(Boolean isUsingDbNull) { _isUsingDbNull = isUsingDbNull; }

    @Override
    public int count(){
        return _data.size();
    }
    @Override
    public void clear(){
        _data.clear();
    }
    @Override
    public boolean exists(String name){
        return _data.containsKey(name);
    }

    @Override
    public Set<String> keys(){
        return _data.keySet();
    }

    @Override
    public DataItem set(String name,Object value)
    {
        _data.put(name, value);
        return this;
    }

    @Override
    public DataItem setIf(boolean condition, String name, Object value) {
        if (condition) {
            _data.put(name, value);
        }
        return this;
    }

    @Override
    public Object get(int index) {
        for (String key : _data.keySet()) {
            if (index == 0) {
                return get(key);
            } else {
                index--;
            }
        }
        return null;
    }

    @Override
    public Object get(String name){
        return _data.get(name);
    }
    @Override
    public Variate getVariate(String name)
    {
        if (_data.containsKey(name)) {
            return new Variate(name, get(name));
        }
        else {
            return new Variate(name, null);
        }
    }

    @Override
    public void remove(String name){
        _data.remove(name);
    }

    @Override
    public <T extends IBinder> T toItem(T item)
    {
        item.bind((key) -> getVariate(key));

        return item;
    }

    @Override
    public short getShort(String name){
        return (short)get(name);
    }

    @Override
    public int getInt(String name){
        return ((Number)get(name)).intValue();
    }

    @Override
    public long getLong(String name){
        return ((Number)get(name)).longValue();
    }

    @Override
    public double getDouble(String name){
        return ((Number)get(name)).doubleValue();
    }

    @Override
    public float getFloat(String name){
        return ((Number)get(name)).floatValue();
    }

    @Override
    public String getString(String name){
        return (String)get(name);
    }

    @Override
    public boolean getBoolean(String name){
        return (boolean)get(name);
    }

    @Override
    public Date getDateTime(String name){
        return (Date)get(name);
    }

    @Override
    public void forEach(Act2<String, Object> callback)
    {
        for(Map.Entry<String,Object> kv : _data.entrySet()){
            Object val = kv.getValue();

            if(val == null && _isUsingDbNull){
                callback.run(kv.getKey(), "$NULL");
            }else {
                callback.run(kv.getKey(), val);
            }
        }
    }

    private boolean _isUsingDbNull=false;

    //============================
    public static IDataItem create(IDataItem schema, GetHandler source) {
        DataItem item = new DataItem();
        for (String key : schema.keys()) {
            Object val = source.get(key);
            if (val != null) {
                item.set(key, val);
            }
        }
        return item;
    }

    @Override
    public Iterator<Map.Entry<String, Object>> iterator() {
        return _data.entrySet().iterator();
    }

    @Override
    public void forEach(Consumer<? super Map.Entry<String, Object>> action) {
        Objects.requireNonNull(action);
        for (Map.Entry<String, Object> entry : _data.entrySet()) {
            action.accept(entry);
        }
    }

    @Override
    public Spliterator<Map.Entry<String, Object>> spliterator() {
        return _data.entrySet().spliterator();
    }


    /** 从map加载数据 */
    public DataItem setMap(Map<String,Object> data) {
        //
        //保持也where的相同逻辑
        //
        return setMapIf(data, (k, v) -> v != null);
    }

    public DataItem setMapIf(Map<String,Object> data, Fun2<Boolean,String,Object> condition) {
        data.forEach((k, v) -> {
            if (condition.run(k, v)) {
                set(k, v);
            }
        });

        return this;
    }

    /** 从Entity 加载数据 */
    public DataItem setEntity(Object obj) {
        //
        //保持也where的相同逻辑
        //
        return setEntityIf(obj, (k, v) -> v != null);
    }

    public DataItem setEntityIf(Object obj, Fun2<Boolean,String,Object> condition) {
        EntityUtils.fromEntity(obj, (k, v) -> {
            if (condition.run(k, v)) {
                set(k, v);
            }
        });
        return this;
    }

    /** 获取map */
    public Map<String,Object> getMap(){
        return _data;
    }


    /**
     *  从Entity 加载数据
     *
     *  可改用：setEntity
     *  */
    @Deprecated
    public void fromEntity(Object obj)  {
        EntityUtils.fromEntity(obj,(k, v)->{
            set(k, v);
        });
    }

    /** 转为Entity */
    public  <T> T toEntity(Class<T> cls) {
        return EntityUtils.toEntity(cls,this);
    }
}

